### PR TITLE
add Lease

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
Signed-off-by: zhangjinpeng1987 <zhangjinpeng@pingcap.com>

### CODE REVIEW

**Comment by zhangjinpeng1987** [comment]
@coocood PTAL

File: tikv/raftstore/util.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+type LeaseState int
+
+const (
+	LeaseState_Suspect LeaseState = 1
+	LeaseState_Valid
+	LeaseState_Expired
+)
+
+type Lease struct {
```
Comment (coocood) [inline tikv/raftstore/util.go]: It's better to port the comments too.

File: tikv/raftstore/util.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+}
+
+func (l *Lease) nextExpiredTime(sendTs time.Time) time.Time {
+	return sendTs.Add(l.maxLease)
+}
+
+func (l *Lease) Renew(sendTs time.Time) {
+	bound := l.nextExpiredTime(sendTs)
+	if l.boundSuspect != nil || l.boundValid != nil {
+		if l.boundSuspect != nil && l.boundSuspect < bound {
```
Comment (coocood) [inline tikv/raftstore/util.go]: time cannot be compared with `<`.

**Comment by coocood** [comment]
We need to port the test for lease too.

File: tikv/raftstore/util.go (hunk 1 above)
Comment (zhangjinpeng1987) [reply tikv/raftstore/util.go]: Done

**Comment by siddontang** [comment]
seem Go time.Now is just using monotonic, we should use monotonic raw here.

**Comment by zhangjinpeng1987** [comment]
@siddontang I don't think it matters, in TiKV we using monotonic raw to fix time jump back issue which is rarely happen.

**Comment by ngaut** [comment]
Using Go native API would be much more easier. We can add a TODO here.

File: tikv/raftstore/util.go [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+/// Renew the lease to the bound.
+func (l *Lease) Renew(sendTs time.Time) {
+	bound := l.nextExpiredTime(sendTs)
+	if l.boundSuspect != nil || l.boundValid != nil {
+		// Longer than suspect ts or longer than valid ts.
+		if l.boundSuspect != nil && l.boundSuspect.Before(bound) {
+			l.boundSuspect = nil
+			l.boundValid = &bound
+		}
+		if l.boundValid != nil && l.boundValid.Before(bound) {
```
Comment (coocood) [inline tikv/raftstore/util.go]: If line 88 set the `boundValid`, then `l.boundValid` would not be nil, it's not the logic we want.
I think this match the orginal logic better.
```go
if l.boundSuspect != nil {
    ...
} else if l.boundValid != nil {
    ...
} else {
    ...
}



```

File: tikv/raftstore/util.go [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+}
+
+const (
+	NSEC_PER_MSEC uint64 = 1000000
+	SEC_SHIFT uint64 = 10
+	MSEC_MASK uint64 = (1 << SEC_SHIFT) - 1
+)
+
+func TimeToU64(t time.Time) uint64 {
+	sec := uint64(t.Unix())
```
Comment (coocood) [inline tikv/raftstore/util.go]: We can use `t.UnixNano()`.

File: tikv/raftstore/util.go [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+	msec := uint64(t.Nanosecond()) / NSEC_PER_MSEC
+	sec <<= SEC_SHIFT
+	return sec | msec
+}
+
+func U64ToTime(u uint64) time.Time {
+	sec := u >> SEC_SHIFT
```
Comment (coocood) [inline tikv/raftstore/util.go]: If use `t.UnixNano` this should be
```
time.Unix(u/1e9, u%1e9)
```

File: tikv/raftstore/util.go (hunk 4 above)
Comment (zhangjinpeng1987) [reply tikv/raftstore/util.go]: I think it should be the nanoseconds in current seconds.

File: tikv/raftstore/util.go (hunk 5 above)
Comment (zhangjinpeng1987) [reply tikv/raftstore/util.go]: Ditto

File: tikv/raftstore/util.go (hunk 4 above)
Comment (coocood) [reply tikv/raftstore/util.go]: In Go source
```
func (t Time) UnixNano() int64 {
	return (t.unixSec())*1e9 + int64(t.nsec())
}
```

**Comment by coocood** [comment]
LGTM

File: tikv/raftstore/util.go (hunk 4 above)
Comment (zhangjinpeng1987) [reply tikv/raftstore/util.go]: What i meant is i need the nanosecond offset within current second.

**Pull request merged**